### ISSUE TITLE
docker compose without "-d" option cannot start container

### ISSUE THREAD

**Issue by emlin** [vmware/vic#4223]
As a user of VIC and docker compose, I should be able to bring up a compose application in the foreground.

Acceptance Criteria:  Bring up a compose application with "docker-compose up" instead of "docker-compose up -d".  If testing this is possible, we should write a robot script to verify this.

---

docker compose does not work without option "-d"
```
docker-compose -H 10.160.63.122:2375 -f link-compose.yml up
Creating network "vic_default" with the default driver
Creating redis

ERROR: for redis  HTTPConnectionPool(host='10.160.63.122', port=2375): Read timed out. (read timeout=60)
ERROR: An HTTP request took too long to complete. Retry with --verbose to obtain debug information.
If you encounter this issue regularly because of slow network conditions, consider setting COMPOSE_HTTP_TIMEOUT to a higher value (current value: 60).
```
The timeout doesn't matter, because there already has error in portlayer log:
```
8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713
Mar 10 2017 21:13:19.569Z ERROR attach connector: Connection not found error for id:8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713: context deadline exceeded
Mar 10 2017 21:13:19.569Z DEBUG [ END ] [github.com/vmware/vic/lib/apiservers/portlayer/restapi/handlers.(*InteractionHandlersImpl).ContainerGetStderrHandler:309] [30.037339079s] 8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713
Mar 10 2017 21:13:19.570Z DEBUG attach connector:  Giving up on connection for 8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713
Mar 10 2017 21:15:53.031Z ERROR Property collector error: Post https://10.160.49.2/sdk: context deadline exceeded
Mar 10 2017 21:15:53.038Z ERROR Unable to wait for extra config property guestinfo.vice..sessions|8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713.started: Post https://10.160.49.2/sdk: context deadline exceeded
Mar 10 2017 21:15:53.038Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).start:166] [3m2.457416913s] 8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713
Mar 10 2017 21:15:53.038Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*Container).start:340] [3m2.457618881s] 8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713
Mar 10 2017 21:15:53.038Z DEBUG [BEGIN] [github.com/vmware/vic/lib/portlayer/exec.(*Container).RefreshFromHandle:323] 8077695f0d45d88e3c2abc26fc201e24
Mar 10 2017 21:15:53.038Z DEBUG Current ChangeVersion: 2017-03-10T21:12:53.727346Z
Mar 10 2017 21:15:53.038Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*Container).RefreshFromHandle:323] [23.267µs] 8077695f0d45d88e3c2abc26fc201e24
Mar 10 2017 21:15:53.038Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.Commit:35] [3m3.398218607s] 8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713
Mar 10 2017 21:15:53.038Z ERROR CommitHandler error on handle(8077695f0d45d88e3c2abc26fc201e24) for 8a8203bbd06c3c061b9925ea682f2b3160da3ce159d25762b07bc7b457368713: &errors.errorString{s:"unable to wait for process launch status: Post https://10.160.49.2/sdk: context deadline exceeded"}
```
Then container stay in Starting status forever
```
docker -H 10.160.63.122:2375  ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS              PORTS                           NAMES
8a8203bbd06c        redis:alpine        "docker-entrypoint..."   10 minutes ago      Starting            10.160.63.122:32770->6379/tcp   redis
```
The thing is suspicious is that portlayer tried to connect container, and container also tried to connect portlayer, but the serial port is not connected.
The difference of docker-compose and general docker command in this case, is that the attach attributes are false from docker-compose.
https://github.com/vmware/vic/blob/master/lib/portlayer/attach/attach.go#L84 is the line we do not actually connect serial port.

Didn't do further investigation for this issue, here is the log bundle, and it's reproducible every time.
[container-logs.zip](https://github.com/vmware/vic/files/835150/container-logs.zip)

**Comment by malikkal** [comment]
Even docker a run without -d doesn't work. I have the same error while trying to interact with container via -it. Even a case opened with support and no resolution yet. Steps to reproduce.

1. Deploy a VCH with --insecure-registry and container network.
2. Run a container out of the registry with -it. In this case a Redhat 6.8 base image.
3. Container will be stuck at Starting...
4. However, the container will run without issues when deployed via Admiral or with -itd.

**Comment by sflxn** [comment]
I'm able to reproduce this.  It appears be a tether related issue.

**Comment by sflxn** [comment]
This is a really nasty bug.  The container started by compose in this way never fully gets into "running" state and cannot be deleted.

docker-compose down fails
docker rm fails

**Comment by sflxn** [comment]
Wrote up #4269 to address the issue in my last comment.

**Comment by sflxn** [comment]
docker-compose up, without -d causes an attach to occur; however, if we inspect the container that's created, we see the following,

```
        "Config": {
            "Hostname": "ef1c9c67e525",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "6379/tcp": {}
            },
```

**Comment by sflxn** [comment]
It turns out the above comment about inspect is the reason this issue occurs.  These flags are used in the portlayer to determine whether to connect the serial port.  Later when compose does an attach, it cannot do so because there is no serial port connected on start.

The fact that compose creates a container with these attach values set to false and then later tries to attach goes against everything we had assumed about these flags.  Our assumption was that you cannot attach to a container if these flags are false on create.  Even the standard docker CLI sets AttachStdout and AttachStderr to true by default.

Either our assumptions about these flags are false or Compose is exploiting a loophole in the daemon that Docker was not aware of at the time.  We still need to find a solution.

**Comment by sflxn** [comment]
At the top of my head, we can resolve this by forcing AttachStdout and AttachStdin to true on ContainerCreate or move the attaching of the serial port to the attach call instead of the create call.

**Comment by sflxn** [comment]
I just created a special build of the docker daemon where I set all the attach flags to false during create.  Using this daemon, I created a container and inspected it to make sure the flags are all set to false.  They are.  I then was able to attach to the container.

The conclusion is that these flags during container create have no meaning for standard docker.  I suspect the only value they have is for regurgitation during docker inspect.

In summary, we can either move the serial port connection to the attach code or we can always attach the serial port on connect, regardless of the attach values.

**Comment by stuclem** [comment]
@sflxn does this need to be included in the release notes for 0.9?

**Comment by hickeng** [comment]
@sflxn the serial port connection code is in the attach path currently - however if stdin/stdout/stderr are all false then attach capability is disabled for that process as it's interpreted to mean that attach is not allowed.
If these fields are simply to record which streams are currently attached then we already have that data from live sources due to our stream handling logic - which io.Copy's are active - so we can use that to populate the fields in the personality.

I'd prefer to keep the current meaning of actual permission in the portlayer as it's a legitimate desire to restrict things like stdin in a prod environment. There's other logged control channels available if needed for controlling apps.

**Comment by sflxn** [comment]
Interesting side effect of enabling attach during create,

docker-compose up     //succeeds
ctrl-c
docker-compose up     //hangs on attach

**Comment by sflxn** [comment]
Also found that inspect data for compose services are incorrect.  Wrote up https://github.com/vmware/vic/issues/4298 for it.

**Comment by stuclem** [comment]
Added the following to the release notes for 0.9:

-------

- <b>Containers remain in the `Starting` state when you run `docker compose up`. [#4223](https://github.com/vmware/vic/issues/4223)</b><br>Bringing up a compose application by using `docker-compose up` results in containers getting stuck at the `Starting` state. This does not happen if you use the `-d` option when you run `docker compose up`.

   Workaround: Use `docker compose up -d` instead of `docker compose up`.

-------

@emlin and @sflxn does this cover it?

**Comment by sflxn** [comment]
Based on feedback on the interaction part for the reattach logic, I've removed that code as it's more involved and touches the attach server.  I'm leaving that part to Caglar since he's currently working on that code right now.  The PR submitted should still handle compose up, without -d.  Reattach for both docker and compose will fail till Caglar gets that fixed.

**Comment by stuclem** [comment]
Already in RNs, as a resolved issue in 1.1.
